From: Chad Monroe <chad@monroe.io>
Date: Fri, 15 May 2026 13:37:37 -0700
Subject: [PATCH] net: ethernet: mtk_eth_soc: expand QDMA TX queue count to 32

PPPQ partitions queues into data and ACK halves (offset 16 on netsys v2+),
which the 16-queue limit cannot accommodate. Raise MTK_QDMA_NUM_QUEUES
to 32 and route every per-queue access through mtk_qdma_select_page() +
MTK_QTX_REG_OFS() so the qdma.page register is paged correctly for
qids 16..31. Use mtk_qtx_sch_is_v3_layout() to pick MIN_RATE field
width.

Also fix the qdma_sched debugfs reader to index by 'i' instead of 'id'
so page 1 is not read through page 0.

Based-on-patch-by: chak-kei.lam <chak-kei.lam@mediatek.com>
Signed-off-by: Chad Monroe <chad@monroe.io>
---
 drivers/net/ethernet/mediatek/mtk_eth_dbg.c |    2 +-
 drivers/net/ethernet/mediatek/mtk_eth_soc.c |    7 +++++--
 drivers/net/ethernet/mediatek/mtk_eth_soc.h |    2 +-
 3 files changed, 7 insertions(+), 4 deletions(-)

--- a/drivers/net/ethernet/mediatek/mtk_eth_dbg.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_dbg.c
@@ -158,9 +158,9 @@ static ssize_t qdma_sched_read(struct fi
 			 (scheduling == 1) ? "WRR" : "SP", max_rate);
 
 	for (i = 0; i < MTK_QDMA_NUM_QUEUES; i++) {
-		mtk_w32(eth, (i / MTK_QTX_PER_PAGE), soc->reg_map->qdma.page);
+		mtk_qdma_select_page(eth, i);
 		sch_reg = mtk_r32(eth, soc->reg_map->qdma.qtx_sch +
-				       (id % MTK_QTX_PER_PAGE) * MTK_QTX_OFFSET);
+				       MTK_QTX_REG_OFS(i));
 		if (mtk_is_netsys_v2_or_greater(eth))
 			scheduler = FIELD_GET(MTK_QTX_SCH_TX_SEL_V2, sch_reg);
 		else
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -825,7 +825,7 @@ static void mtk_set_queue_speed(struct m
 				int speed)
 {
 	const struct mtk_soc_data *soc = eth->soc;
-	u32 ofs, val;
+	u32 val;
 
 	if (!MTK_HAS_CAPS(soc->caps, MTK_QDMA))
 		return;
@@ -916,8 +916,8 @@ static void mtk_set_queue_speed(struct m
 		}
 	}
 
-	ofs = MTK_QTX_OFFSET * idx;
-	mtk_w32(eth, val, soc->reg_map->qdma.qtx_sch + ofs);
+	mtk_qdma_select_page(eth, idx);
+	mtk_w32(eth, val, soc->reg_map->qdma.qtx_sch + MTK_QTX_REG_OFS(idx));
 }
 
 static void mtk_gdm_mac_link_up(struct mtk_mac *mac,
@@ -2872,7 +2872,7 @@ static int mtk_tx_alloc(struct mtk_eth *
 	int i, sz = TX_DESC_OFS(eth, 1);
 	struct mtk_tx_dma_v2 *txd;
 	int ring_size;
-	u32 ofs, val;
+	u32 val;
 
 	if (MTK_HAS_CAPS(soc->caps, MTK_QDMA))
 		ring_size = MTK_QDMA_RING_SIZE;
@@ -2947,25 +2947,26 @@ static int mtk_tx_alloc(struct mtk_eth *
 			soc->reg_map->qdma.crx_ptr);
 		mtk_w32(eth, ring->last_free_ptr, soc->reg_map->qdma.drx_ptr);
 
-		for (i = 0, ofs = 0; i < MTK_QDMA_NUM_QUEUES; i++) {
+		for (i = 0; i < MTK_QDMA_NUM_QUEUES; i++) {
+			mtk_qdma_select_page(eth, i);
+
 			val = (QDMA_RES_THRES << 8) | QDMA_RES_THRES;
-			mtk_w32(eth, val, soc->reg_map->qdma.qtx_cfg + ofs);
+			mtk_w32(eth, val,
+				soc->reg_map->qdma.qtx_cfg + MTK_QTX_REG_OFS(i));
 
 			val = MTK_QTX_SCH_MIN_RATE_EN |
 			      MTK_QTX_SCH_LEAKY_BUCKET_SIZE;
 			/* minimum: 10 Mbps */
-			if (mtk_is_netsys_v3_or_greater(eth) &&
-			    (eth->soc->caps != MT7988_CAPS)) {
+			if (mtk_qtx_sch_is_v3_layout(eth))
 				val |= FIELD_PREP(MTK_QTX_SCH_MIN_RATE_MAN_V3, 1) |
 				       FIELD_PREP(MTK_QTX_SCH_MIN_RATE_EXP_V3, 4);
-			} else {
+			else
 				val |= FIELD_PREP(MTK_QTX_SCH_MIN_RATE_MAN, 1) |
 				       FIELD_PREP(MTK_QTX_SCH_MIN_RATE_EXP, 4);
-			}
 			if (mtk_is_netsys_v1(eth))
 				val |= MTK_QTX_SCH_LEAKY_BUCKET_EN;
-			mtk_w32(eth, val, soc->reg_map->qdma.qtx_sch + ofs);
-			ofs += MTK_QTX_OFFSET;
+			mtk_w32(eth, val,
+				soc->reg_map->qdma.qtx_sch + MTK_QTX_REG_OFS(i));
 		}
 		val = MTK_QDMA_TX_SCH_MAX_WFQ | (MTK_QDMA_TX_SCH_MAX_WFQ << 16);
 		mtk_w32(eth, val, soc->reg_map->qdma.tx_sch_rate);
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -26,7 +26,7 @@
 #define MTK_MAX_DSA_PORTS	7
 #define MTK_DSA_PORT_MASK	GENMASK(2, 0)
 
-#define MTK_QDMA_NUM_QUEUES	16
+#define MTK_QDMA_NUM_QUEUES	32
 #define MTK_QDMA_PAGE_SIZE	2048
 #define MTK_MAX_RX_LENGTH	1536
 #define MTK_MAX_RX_LENGTH_2K	2048
